notebooks/Misconfiguration_detection_framework_for_data_type_errors.ipynb

```diff
@@ -0,0 +1,743 @@
+{
```

Comment:
<p>"Percentage of data type errors in mysql.cnf files are 20.47244094488189 " Can this be validated in some way? Correct me if I am wrong, but the meaning here is, "according to the rules learned above, 20% of the configurations are using the incorrect data type". Are we able to confirm the rules being learned are accurate? 20% configuration errors seems high.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/configuration-files-analysis/pull/2/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Reply:
<p>Sanket, did we labels for this data? If not, can you give examples of the data type errors here? I'm guessing there would be repetitive errors so maybe  a frequency table. </p><p>Also, you could round up the floats.</p>

Reply:
I am in the process of making changes and will update shortly.

Reply:
Hi @MichaelClifford , @Shreyanand , 

Thanks for the feedback. 

1) There is a bug in the insight configuration parser. If we have a file with inline comments, the parser is not working as expected. I have created an issue for the same. (https://github.com/RedHatInsights/insights-core/issues/2777)
2) Certain conditions for assigning data types are strict. Such as if data type is size i.e. 14MB, 1G etc. it didnt accept the integer values. If most of the values are 0||1 consider the data type as 0||1; It will create more warnings(FP), but it might catch the error. 
**I relaxed all these kinds of conditions in this commit.**  
3) Finally, I found it's difficult to include all the conditions for determining the data type from given values using RegEx. Like filename [conventions](https://en.wikipedia.org/wiki/Filename), a file-path convention for different OS(in this data many configuration files are configured for windows). There are always other corner cases for determining data types such as urls, IP address, ipnetwork etc.

Hence, I removed all the regex base data inference techniques except for size determination as it is a simple regex query. I replaced all the techniques by using python libraries such as ipaddress, pathvalidatior, urllib etc. 
I made a certain modification in those functions according to our needs. 


> Sanket, did we labels for this data?

No, this data is not labeled i.e we don't know it is type error or not. Basically, all the files are collected from the web, including Q&A forums and the official mailing lists of the software such as StackOverflow. many time issues arise because users manually anonymized the path of files, port numbers etc. Most of the urls are not anymore valid and we just know the one word [issue](https://github.com/tianyin/configuration_datasets) 

After making all these changes I mentioned above, I validated all the rules and detected errors by myself. All of them are making sense. Percentage of errors are currently 10.63 most and most of the errors were are fixed because of point number 2: many time users provided size in integer format. 10-15 files.

currently, most of the detected issues is caused by the bug in the insight parser that I mentioned in point number 2. Hence, I am not sure is it fine to show all the errors in the notebook.